runtime/protocol/http-client-engines/http-client-engine-crt/jvm/src/aws/smithy/kotlin/runtime/http/engine/crt/ConnectionManager.kt

```diff
@@ -82,8 +93,18 @@ internal class ConnectionManager(
             }
 
             throw httpEx
+        } finally {
+            metrics.queuedRequests = pending.decrementAndGet()
+            emitConnections()
         }
     }
+
+    private fun emitConnections() {
+        val idleConnections = leases.availablePermits.toLong()
```

Comment:
**correctness**: this semaphore isn't tied in anyway to actual connections in CRT, you could have 100 max connections configured but that doesn't mean you have 100 idle connections already connected waiting to be used. I'm thinking this would have to come from the actual CRT connection manager.

Reply:
Oh jeez, good point. Yes, this is clearly flawed.

That brings up an interesting wrinkle, though. We don't have just one CRT connection manager, we have [one for each host](https://github.com/awslabs/smithy-kotlin/blob/49a0089293068cdb9aec6d07fe019080c3b14117/runtime/protocol/http-client-engines/http-client-engine-crt/jvm/src/aws/smithy/kotlin/runtime/http/engine/crt/ConnectionManager.kt#L55). Does the Smithy metric for acquired connections equal the sum of acquired connections _across all_ underlying CRT connection managers?

Moreover, we have [a CRT config param called `maxConnections`](https://github.com/awslabs/smithy-kotlin/blob/49a0089293068cdb9aec6d07fe019080c3b14117/runtime/protocol/http-client-engines/http-client-engine-crt/jvm/src/aws/smithy/kotlin/runtime/http/engine/crt/CrtHttpEngineConfig.kt#L31-L34). Contrary to its name, it's [used as the maximum number of connections _for each_ underlying CRT connection manager](https://github.com/awslabs/smithy-kotlin/blob/49a0089293068cdb9aec6d07fe019080c3b14117/runtime/protocol/http-client-engines/http-client-engine-crt/jvm/src/aws/smithy/kotlin/runtime/http/engine/crt/ConnectionManager.kt#L50). Meaning if `maxConnections` is 100 and we connect to 3 hosts, there are actually 3 underlying CRT connection managers which _each_ have a max of 100 connections for a total of 300. How then do we measure the Smithy metric of idle connections? Is it the sum of idle connections _across all_ underlying CRT connection managers (which may be more than `maxConnections`)? Or is it `maxConnections` minus the sum of acquired connections _across all_ underlying CRT connection managers (which may be less than the actual number of open/idle connections)?

Reply:
The fact CRT uses connection manager per host is an implementation detail, the config setting is at the engine level. This is what the `leases` semaphore is for is to control max connections across multiple managers (albeit it likely isn't perfect at ensuring we never cross the `maxConnections` threshold due to idle connections).

The reason I allowed each connection manager to be configured with `maxConnections` is because we don't know how many hosts an engine will be used to connect to, it may be 1 or it may be split across many hosts which will require many connection managers.

Reply:
I see this as an area that may cause confusion later, especially if users begin correlating the idle connections metric with connections reported by the OS or JVM. I don't have a better suggestion though given how the CRT connection manager works. I'll switch the implementation to calculate the idle connections metric as `maxConnections` minus the sum of acquired connections across all underlying CRT connection managers.

Reply:
Oh and interrogating the number of active connections from the CRT connection managers requires another revision of **aws-crt-kotlin**: https://github.com/awslabs/aws-crt-kotlin/pull/88